services/src/main/scala/cromwell/pipeline/service/ProjectSearchService.scala

```diff
@@ -0,0 +1,114 @@
+
+trait ProjectSearchService {
+
+  def searchProjects(request: ProjectSearchRequest, userId: UserId): Future[ProjectSearchResponse]
+
+}
+
+object ProjectSearchService {
+
+  object Exceptions {
+    sealed abstract class ProjectSearchException(message: String) extends ServiceException(message)
+    final case class InternalError(message: String = "Internal error") extends ProjectSearchException(message)
+  }
+
+  def apply(
+    projectService: ProjectService,
+    projectFileService: ProjectFileService,
+    projectConfigurationService: ProjectConfigurationService
+  )(
+    implicit ec: ExecutionContext
+  ): ProjectSearchService =
+    new ProjectSearchService {
+
+      override def searchProjects(request: ProjectSearchRequest, userId: UserId): Future[ProjectSearchResponse] = {
+        val userProjects = projectService.getUserProjects(userId).recoverWith {
+          case _ => internalError("fetch projects")
+        }
+
+        def doSearch(searchFilter: ProjectSearchFilter, projects: Future[Seq[Project]]): Future[Seq[Project]] =
+          searchFilter match {
+            case All                 => userProjects
+            case ByName(mode, value) => userProjects.map(projects => searchByNameFilter(mode, value, projects))
+            case ByFiles(_, hasFiles) =>
+              userProjects.flatMap(projects => searchByFilesFilter(hasFiles, projects, userId))
+            case ByConfig(_, hasConfig) =>
+              userProjects.flatMap(projects => searchByConfigFilter(hasConfig, projects, userId))
+            case Or(leftFilter, rightFilter) =>
+              doSearch(leftFilter, projects).flatMap(
+                resultLeft => doSearch(rightFilter, projects).map(resultRight => (resultLeft ++ resultRight).distinct)
+              )
+            case And(leftFilter, rightFilter) =>
+              val leftResult = doSearch(leftFilter, projects)
+              doSearch(rightFilter, leftResult)
+          }
+        doSearch(request.filter, userProjects).map(projects => ProjectSearchResponse(projects))
+
+      }
+
+      private def searchByConfigFilter(
+        hasConfig: Boolean,
+        projects: Seq[Project],
+        userId: UserId
+      ): Future[Seq[Project]] = {
+        def getProjectWithConfig(project: Project): Future[(Project, Option[ProjectConfiguration])] =
+          projectConfigurationService.getLastByProjectId(project.projectId, userId).transformWith {
+            case Success(configuration) => Future.successful(project, Some(configuration))
+            case Failure(NotFound(_))   => Future.successful(project, None)
+            case _                      => internalError("fetch configuration")
+          }
+
+        for {
+          projWithConf <- Future.sequence(projects.map(getProjectWithConfig))
+        } yield projWithConf.collect {
+          case (proj, Some(_)) if hasConfig => Some(proj)
+          case (proj, None) if !hasConfig   => Some(proj)
+        }.flatten
```

Comment:
Why do you need to wrap all possible cases to `Some(_)` and then flatten it? Can't you just take result as is?
```suggestion
        } yield projWithConf.collect {
          case (proj, Some(_)) if hasConfig => proj
          case (proj, None) if !hasConfig   => proj
        }
```
Please have a look at documentation on how to work with `collect` method. Same applies to `searchByFilesFilter`

Reply:
Fixed